proto/v3/skaffold.proto

```diff
@@ -0,0 +1,549 @@
+}
+
+// TestMetadata describes the test pipeline
+message TestMetadata {
+    message Tester {
+        enums.TesterType type = 1;
+        int32 count = 2;
+    }
+    repeated Tester Testers = 1;
+}
+
+
+// RenderMetadata describes the render pipeline
+message RenderMetadata {
+    message Renderer {
+        enums.RenderType type = 1;
+        int32 count = 2;
+    }
+    repeated Renderer Renderers = 1;
+}
+
+message DeployMetadata {
+    message Deployer {
+        enums.DeployerType type = 1;
+        int32 count = 2;
+    }
+    repeated Deployer deployers = 1;
+    enums.ClusterType cluster = 2;
+}
+
+// `BuildState` maps Skaffold artifacts to their current build states
+message BuildState {
+    // A map of `artifact name -> build-state`.
+    // Artifact name is defined in the `skaffold.yaml`.
+    // The `build-state` can be: <br>
+    // - `"NotStarted"`: not yet started <br>
+    // - `"InProgress"`: build started <br>
+    // - `"Complete"`: build succeeded <br>
+    // - `"Failed"`: build failed
+    map<string, string> artifacts = 1;
+    bool autoTrigger = 2;
+    enums.StatusCode statusCode = 3;
+}
+
+// `TestState` describes the current state of the test
+message TestState {
+    // Status of the current test
+    string status = 1;
+    // Teststate status code
+    enums.StatusCode statusCode = 2;
+}
+
+// `RenderState` describes the current state of the render
+message RenderState {
+    // Status of the current render
+    string status = 1;
+    // Renderstate status code
+    enums.StatusCode statusCode = 2;
+}
+
+// `DeployState` describes the status of the current deploy
+message DeployState {
+    string status = 1;
+    bool autoTrigger = 2;
+    enums.StatusCode statusCode = 3;
+}
+
+// `StatusCheckState` describes the state of status check of current deployed resources.
+message StatusCheckState {
+    string status = 1;
+    // A map of `resource name -> status-check-state`. Where `resource-name` is the kubernetes resource name.
+    // The `status-check-state` can be <br>
+    // - `"NotStarted"`: indicates that `status-check` has just started. <br>
+    // - `"InProgress"`: InProgress is sent after every resource check is complete. <br>
+    // - `"Succeeded"`:
+    // - `"Failed"`:
+    map<string, string> resources = 2;
+    // StatusCheck statusCode
+    enums.StatusCode statusCode = 3;
+
+}
+
+// `FileSyncState` contains the status of the current file sync
+message FileSyncState {
+    string status = 1;
+    bool autoTrigger = 2;
+}
+
+// `Event` describes an event in the Skaffold process.
+// It is a wrapper on the top of other events like MetaEvent, BuildEvent, TestEvent, DeployEvent, PortEvent, StatusCheckEvent, ResourceStatusCheckEvent, FileSyncEvent, or DebuggingContainerEvent.
+message Event {
+    string specversion = 1;
+    string type = 2;
+    string source = 3;
+    string id = 4;
+    google.protobuf.Timestamp time = 5;
+    google.protobuf.Any data = 6;
+    string datacontenttype = 7;
+}
```

Comment:
@briandealwis We are following [CloudEvents Spec](https://github.com/cloudevents/spec) and not [CloudEvents Protobuf Spec](https://github.com/cloudevents/spec/blob/v1.0.1/protobuf-format.md).
Reason for not going with CloudEvent Protobuf spec - Not all major Language's SDK support CloudEvent protobuf specs. For instance, while [Java SDK](https://cloudevents.github.io/sdk-java/protobuf.html) has support for this but not [Golang](https://github.com/cloudevents/sdk-go)

Reply:
The CE Go SDK [does support protobuf format](https://github.com/cloudevents/sdk-go/blob/236e4413fbc4508b54cf5c45826adf2c4af61376/binding/format/protobuf/v2/protobuf.go) but it's undocumented.

That said, I don't know how this would integrate into grpc-framework.